Avoid possible filename collision in coverage tests

Swatinem opened the pull request:
Previously, coverage tests were writing profiler data to files based on
their pid. As rustdoc spawns each doctest as its own process, it might
be possible in rare cases that a pid is being reused which would cause
a file to be overwritten, leading to incorrect coverage results.

should help with #83262

r? @tmandry

Swatinem commented on src/test/run-make-fulldeps/coverage-reports/Makefile:
```diff
@@ -106,6 +106,7 @@ endif
 	# Postprocess the profiling data so it can be used by the llvm-cov tool
 	"$(LLVM_BIN_DIR)"/llvm-profdata merge --sparse \
+			"$(TMPDIR)"/$@.profraw \
 			"$(TMPDIR)"/$@-*.profraw \
```
Not sure if this make/shell syntax will fail if no such files exist? In that case I might have to replicate the shell-for from down below.

TBH, I haven’t run this yet, its quite a pain to make this work on windows :-(

richkadel commented:
@Swatinem - Thanks for the PR.

I need more time to consider these changes than I have right now (on vacation).

Would you be willing to upload a separate change that disables the doctest coverage test on MSVC (as discussed in https://github.com/rust-lang/rust/issues/83262#issuecomment-841385944) in the mean time?

If you feel strongly about this change, you could ask @tmandry or @wesleywiser to take a look in my place. Otherwise, I'll take a look at this PR next week.

Swatinem commented:
ping @richkadel might this still be a worthwhile change for right now?

Surely, refactoring the tests to run directly via rustbuild/compiletest might be interesting, but not really something I feel like getting into right now.

richkadel commented:
Sorry for the delay! I'll take a look now.

richkadel commented on src/test/run-make-fulldeps/coverage-reports/Makefile:
```diff
@@ -87,7 +87,7 @@ endif
 	# with `LLVM_PROFILE_FILE=<profdata_file>` environment variable set to
 	# output the coverage stats for this run.
-	LLVM_PROFILE_FILE="$(TMPDIR)"/$@-%p.profraw \
+	LLVM_PROFILE_FILE="$(TMPDIR)"/$@.profraw \
```
👍  I support the change on this line (removing pid from the non-doc-test run), even though it shouldn't make a difference in the result.

richkadel commented on src/test/run-make-fulldeps/coverage-reports/Makefile:
```diff
@@ -98,28 +98,29 @@ endif
 
 	# Run it through rustdoc as well to cover doctests
-	LLVM_PROFILE_FILE="$(TMPDIR)"/$@-%p.profraw \
+	LLVM_PROFILE_FILE="$(TMPDIR)"/$@-%p-%m.profraw \
```
I _think_ this will help, assuming our theory is correct that there could be PID-reuse on Windows. As I understand it, every doc test produces a new, program (new `main()` and unique source) so the binary signature (`%m`) should be different for each one.

In theory, adding the PID `%p` value is probably superfluous, because `%m` _should_ be unique enough, but maybe keeping `%p` is an added safeguard to ensuring we don't have filename collisions, and there's probably harm keeping both.

richkadel commented on src/test/run-make-fulldeps/coverage-reports/Makefile:
```diff
@@ -98,28 +98,29 @@ endif
 	# Postprocess the profiling data so it can be used by the llvm-cov tool
 	"$(LLVM_BIN_DIR)"/llvm-profdata merge --sparse \
-			"$(TMPDIR)"/$@-*.profraw \
+			--num-threads=1 \
```
Based on the documentation, and just googling for more info on the `--num-threads` option to both `llvm-profdata` and `llvm-cov`, I don't see how this option would have any impact.

For example, https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/llvm-cov/multithreaded-report.test seems to imply the option should have no effect on the results.

If you have evidence of some undocumented side effect of adding this option, please add a comment in the Makefile explaining the side effect you expect, and a link to the online reference (documentation or code) that shows the evidence of this side effect. 

Otherwise, I don't think this option should be added, in either command.

I

richkadel reviewed (changes requested):
@Swatinem - Thanks for the PR!

I left some inline comments. In my view, the only change that I _think_ has a chance of resolving the issue (if our theory is right about reused PIDs) is adding the `%m`.

Please let me know if you see things differently, and add comments to the Makefile that explain the reasons for options you are adding and the expected impact/benefit of those options (for future reviewers).

Also, I don't understand the PR title, particularly, "Order profraw files explicitly". (I thought this was about filename collisions, and I don't see anything in your change that has to do with ordering files.)

What I think is happening (best guess) is the `.profraw` files are generated for each doc test, and at least some of the doctests are run under the same Windows PID. If that's happening, then some doc test `.profraw` results are overwritten, so the coverage for them is lost.

That would happen before running `llvm-profdata`, so adding `--num-threads` to the LLVM commands doesn't seem relevant to me.

So I may not fully understand what your change is doing, but if you agree with what I've said, then I recommend changing the title to be more accurate, something like, "Address possible PID-based profraw filename collisions on doc tests".

richkadel commented:
r? @tmandry 

(I don't have privileges to approve in rust-lang/rust, but Tyler can.)

richkadel commented:
> r? @tmandry

@Swatinem - You may have to update the reviewer. It doesn't seem like highfive is responding to my `r?` directive.

Swatinem commented:
r? @tmandry 

I rebased and added another comment, removing unrelated changes. (I was trying different things but couldn’t reproduce this) Will change the PR title and description in a moment.

tmandry commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 084794ed164199619d9bd6e5422b4feaee4a0a9a has been approved by `tmandry`

<!-- @bors r=tmandry 084794ed164199619d9bd6e5422b4feaee4a0a9a -->
<!-- homu: {"type":"Approved","sha":"084794ed164199619d9bd6e5422b4feaee4a0a9a","approver":"tmandry"} -->
